hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/recordcache/DeduplicationCacheImpl.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.state.recordcache;
+
+import static com.hedera.node.app.spi.HapiUtils.ACCOUNT_ID_COMPARATOR;
+import static com.hedera.node.app.spi.HapiUtils.TIMESTAMP_COMPARATOR;
+import static com.hedera.node.app.spi.HapiUtils.asTimestamp;
+import static com.hedera.node.app.spi.HapiUtils.minus;
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.hapi.node.base.TransactionID;
+import com.hedera.node.app.service.mono.context.properties.GlobalDynamicProperties;
+import com.hedera.node.app.state.DeduplicationCache;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.time.Instant;
+import java.util.Comparator;
+import java.util.Set;
+import java.util.concurrent.ConcurrentSkipListSet;
+import javax.inject.Inject;
+
+/** An implementation of {@link DeduplicationCache}. */
+public final class DeduplicationCacheImpl implements DeduplicationCache {
+    /**
+     * The {@link TransactionID}s that this node has already submitted to the platform, sorted by transaction start
+     * time, such that earlier start times come first. We guard this data structure within a synchronized block.
+     */
+    private final Set<TransactionID> submittedTxns = new ConcurrentSkipListSet<>(
+            (t1, t2) -> Comparator.comparing(TransactionID::transactionValidStartOrThrow, TIMESTAMP_COMPARATOR)
+                    .thenComparing(TransactionID::accountID, ACCOUNT_ID_COMPARATOR)
+                    .compare(t1, t2));
+
+    /** Used for looking up the max transaction duration window. To be replaced by some new config object */
+    private final GlobalDynamicProperties props;
+
+    /** Constructs a new {@link DeduplicationCacheImpl}. */
+    @Inject
+    public DeduplicationCacheImpl(@NonNull final GlobalDynamicProperties props) {
+        this.props = requireNonNull(props);
```

Comment:
In the future, `Configuration` will be immutable and we need to be able to submit the current version. This will requires some modifications. Maybe the configuration can be passed as an additional parameter to the methods that need the configuration.

Reply:
DeduplicationCache works outside of consensus and isn't deterministic -- can be a little sloppy. I think we could have a supplier that gets the configuration from the latest immutable state. That would be good enough, and could avoid having to pass the configuration into the different methods.

Reply:
OK I have an implementation, will push as soon as tests are done.